Make geo queries first class citizen

iverase opened the pull request:
This PR proposes to create a new method on `MappedFieldType`  called `geoQuery()` as we are doing with other queries. That will allow to easilly integrate the different geo queries over different geo fields.

In addition we need to add another method called `spatialQuery` to support shape queries (spatial queries on cartesian space).

elasticmachine commented:
Pinging @elastic/es-analytics-geo (:Analytics/Geo)

iverase commented on server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:
```diff
@@ -407,6 +408,16 @@ public SpanQuery spanPrefixQuery(String value, SpanMultiTermQueryWrapper.SpanRew
+    public Query spatialQuery(Geometry shape, String fieldName, ShapeRelation relation, QueryShardContext context) {
+        throw new IllegalArgumentException("Can only use shape queries on shape fields - not on [" + name
+            + "] which is of type [" + typeName() + "]");
+    }
```
maybe we should only have one method here? spatialQuery?

iverase commented:
The reason I went this path is that I found many methods in this class that seems to be only implemented by few fields (I think sometimes only by one), for example phrasePrefixQuery or spanPrefixQuery.... I believe can only be implemented by text field.

With this change I propose to dd only one method, maybe two that will be implemented so far by 4 fields: (geo_)point & (geo_)shape.

iverase commented:
@imotov I have introduced a new Field type called `SpatialFieldType` that contains a method to build spatial queries. This actually looks better, thanks!

Note that I love the method that creates a ShapeBuilder from a Geometry to GeoUtils.

imotov commented:
This addresses my first concern. Thanks!

iverase commented:
@imotov let me know if you have time to review this, is getting into the critical path so I can look for someone else.

imotov commented:
Yes, I was waiting for you to address my second concern: 

> by removing the QueryProcessor we convert *ShapeFieldMapper back into a "god class".

iverase commented:
ahh! I must say I did not understand your concern and now looking into more detail I think I disagree that this change makes Shape Mappers 'god classes'.  I think the interface `QueryProcesor` is actually wrong:

* Name is confusing as it is not processing any query, it is building a Lucene query from a Elasticsearch geometry object.

* It increases unnecessary the surface API IMHO. Field types seems to be responsible in generating Lucene queries, therefore let's move that responsibility there, where it belongs.

To alleviate your feeling of `god class` I move the code that generates the Lucene query to the field type by adding the static method `makeQueryFromGeometry`.  Let me know what you think and why you think doing so it makes the Shape mappers a `god class` as I don't get it.

Other things in this iteration:

* I extracted all the methods that generates Lucene geometries from Elasticsearch geometries into an Utils class. 

* Now that we depend on Lucene 8.5, I have changed the way we build the queries. We have a new method where you can pass an array of Lucene geometries and it is Lucene that generates the most efficient query depending on the geometries and the spatial operation. For example, for geometry collections we were creating one query for geometry. Lucene will only generate one query for the whole collection except for CONTAINS where it build one query for geometry.

imotov commented:
> Name is confusing as it is not processing any query, it is building a Lucene query from a Elasticsearch geometry object.

I am ok with renaming it :)

> It increases unnecessary the surface API IMHO. Field types seems to be responsible in generating Lucene queries, therefore let's move that responsibility there, where it belongs.

It implements a simple, well defined interface. 

> I move the code that generates the Lucene query to the field type by adding the static method makeQueryFromGeometry

That splits the logic between Utility class and the mapper, so I think it is the worst solution here. If you cannot live with the processor interface, let's move the whole makeQueryFromGeometry to an utility class. This method has a well defined purpose, with minimal amount of dependencies (QueryShardContext leaks there, but we can remove it by throwing Unsupported operation exception and wrapping it one level above). At least, it will be easily unit testable and we can greatly simplify the test setup for it.

iverase commented:
I am sorry but I do not understand what you want to achieve, The interface is doing exactly what the SpatialFieldType#spatialQuery is suppose to do. I cannot see how to integrate it sorry.

imotov commented:
> I am sorry but I do not understand what you want to achieve,

I want to be able to unit test the functionality inside SpatialFieldType#spatialQuery without instantiating or mocking DocumentMapper. It has non trivial logic independent from the Mapping logic. I think they should be separated if possible. 

For example, if we take a look at [this test](https://github.com/elastic/elasticsearch/blob/a8b39ed842c7770bd9275958c9f747502fd9a3ea/server/src/test/java/org/elasticsearch/index/mapper/LegacyGeoShapeFieldMapperTests.java#L705-L728) lines [708-719](https://github.com/elastic/elasticsearch/blob/a8b39ed842c7770bd9275958c9f747502fd9a3ea/server/src/test/java/org/elasticsearch/index/mapper/LegacyGeoShapeFieldMapperTests.java#L708-L719) with all fake indices and services that we are creating are not needed. I was planning to clean all this stuff up but got busy with other tasks and forgot. I am really sorry about that. If we put all the query generating back inside FieldMappers such clean up would be impossible.

Does it make sense?

iverase commented:
You can always instantiate a GeoShapeFieldMapperType and unit test that? Not need to mock up anything or I am missing something

nknize commented:
Hmm... Before merging, we need to look at how this might impact [Coordinate Reference System support](https://github.com/elastic/elasticsearch/pull/47250/files#diff-1dc62aec3158b0b32fbb78e61afe8b22R113) which relies on the index and query interfaces to override OSS functionality in XPack through the registered DataHandler.

nknize commented:
Is this PR largely needed so `GeoPointFieldType` can implement `public Query spatialQuery(...)` in order to implement `geo_shape` queries on `geo_point` field types?  Can that not be achieved using the queryProcessor interface? (which we'll likely need when overriding by coordinate reference system?)

iverase commented:
Thanks Nick for pointing that PR. Indeed this change will break it and therefore it is a bug, so I am closing without merging.

The first approach was to use the `AbstractGeometryFieldMapper` interface but this is doing much more than we need.  Igor proposed to break into a searchable part, I thought it would make sense to move it to the field type. It turns out this is a bug.

Thanks @imotov for your patience, so I am closing this. In the long term we will probably want `geo_point` to extend fully `AbstractGeometryFieldMapper`, in the short team we can break the field type interface.
